From 80e9caa4d70b7bf853de4ec5a41e74f14f8a93ba Mon Sep 17 00:00:00 2001
From: OpenEmbedded <oe.patch@oe>
Date: Mon, 11 Aug 2025 09:48:18 +0000
Subject: [PATCH 1/2] Workaround mctp-pcie-vdm and mctp route to support
 routing of broadcast message from driver to mctp network stack.

Introduced below solution to fix message forwarding issue.
Credit:
	https://github.com/Kostr/linux/tree/non_zero_instance_id
	https://github.com/AspeedTech-BMC/linux/pull/3

Upstream-Status: Inappropriate [No upstream maintenance]
---
 drivers/net/mctp/mctp-pcie-vdm.c | 23 +++++++++++++++++------
 include/linux/aspeed-mctp.h      |  2 +-
 net/mctp/route.c                 | 11 ++++++++++-
 3 files changed, 28 insertions(+), 8 deletions(-)

diff --git a/drivers/net/mctp/mctp-pcie-vdm.c b/drivers/net/mctp/mctp-pcie-vdm.c
index 4b43e348745f..cc238dd138b3 100644
--- a/drivers/net/mctp/mctp-pcie-vdm.c
+++ b/drivers/net/mctp/mctp-pcie-vdm.c
@@ -92,7 +92,7 @@ enum mctp_ctrl_command_code {
 };
 
 struct mctp_ctrl_msg_hdr {
-	u8 ctrl_msg_class;
+	u8 instanceID : 5, rsvd : 1, datagram : 1, rq : 1;
 	u8 command_code;
 };
 
@@ -198,6 +198,8 @@ static void mctp_pcie_vdm_ctrl_msg_handler(struct mctp_pcie_vdm_dev *vdm_dev,
 	u8 message_type =
 		FIELD_GET(MCTP_MSG_TYPE_MASK, packet[MCTP_PCIE_VDM_HDR_SIZE]);
 
+	pr_debug("%s: Message Type: %d", __func__, message_type);
+
 	if (message_type != MCTP_CONTROL_MSG_TYPE)
 		return;
 
@@ -207,9 +209,10 @@ static void mctp_pcie_vdm_ctrl_msg_handler(struct mctp_pcie_vdm_dev *vdm_dev,
 	/* host endian expected */
 	struct mctp_pcie_vdm_hdr *hdr = (struct mctp_pcie_vdm_hdr *)packet;
 
+	pr_debug("%s: Command Code: %d", __func__, ctrl_hdr->command_code);
 	switch (ctrl_hdr->command_code) {
 	case MCTP_CTRL_CMD_SET_ENDPOINT_ID:
-		if (ctrl_hdr->ctrl_msg_class == MCTP_CTRL_MSG_RQDI_REQ) {
+		if ((ctrl_hdr->rq == 1) && (ctrl_hdr->datagram == 0)) {
 			/* EID placed at byte2 of SET EID REQ DATA */
 			u8 dst_eid =
 				packet[MCTP_PCIE_VDM_HDR_SIZE +
@@ -221,7 +224,7 @@ static void mctp_pcie_vdm_ctrl_msg_handler(struct mctp_pcie_vdm_dev *vdm_dev,
 		}
 		break;
 	case MCTP_CTRL_CMD_GET_ENDPOINT_ID:
-		if (ctrl_hdr->ctrl_msg_class == MCTP_CTRL_MSG_RQDI_RSP) {
+		if ((ctrl_hdr->rq == 0) && (ctrl_hdr->datagram == 0)) {
 			/* EID placed at byte2 of GET EID RSP DATA */
 			u8 target_eid =
 				packet[MCTP_PCIE_VDM_HDR_SIZE +
@@ -235,24 +238,32 @@ static void mctp_pcie_vdm_ctrl_msg_handler(struct mctp_pcie_vdm_dev *vdm_dev,
 	case MCTP_CTRL_CMD_DISCOVERY_NOTIFY:
 		hdr->pci_target_id = 0x0000;
 		/* default use MCTP_PCIE_VDM_ROUTE_BY_ID, so no need to handle RSP class */
-		if (ctrl_hdr->ctrl_msg_class == MCTP_CTRL_MSG_RQDI_REQ) {
+		if ((ctrl_hdr->rq == 1) && (ctrl_hdr->datagram == 0)) {
 			hdr->route_type = MCTP_PCIE_VDM_TYPE_MSG |
 					  MCTP_PCIE_VDM_ROUTE_TO_RC;
 		}
 		break;
 	case MCTP_CTRL_CMD_PREPARE_ENDPOINT_DISCOVERY:
 	case MCTP_CTRL_CMD_ENDPOINT_DISCOVERY:
-		if (ctrl_hdr->ctrl_msg_class == MCTP_CTRL_MSG_RQDI_REQ) {
+		if ((ctrl_hdr->rq == 1) && (ctrl_hdr->datagram == 0)) {
+			pr_debug("%s: MCTP_CTRL_CMD_ENDPOINT_DISCOVERY Req from %02x: %x", __func__,
+                hdr->pci_req_id,
+			    hdr->route_type);
 			hdr->route_type = MCTP_PCIE_VDM_TYPE_MSG |
 					  MCTP_PCIE_VDM_BROADCAST_FROM_RC;
 			hdr->pci_target_id = 0xFFFF;
-		} else if (ctrl_hdr->ctrl_msg_class == MCTP_CTRL_MSG_RQDI_RSP) {
+			u16 src_bdf = hdr->pci_req_id;
+		} else if ((ctrl_hdr->rq == 0) && (ctrl_hdr->datagram == 0)) {
 			hdr->route_type = MCTP_PCIE_VDM_TYPE_MSG |
 					  MCTP_PCIE_VDM_ROUTE_TO_RC;
+			pr_debug("%s: MCTP_CTRL_CMD_ENDPOINT_DISCOVERY Resp: %x", __func__,
+			    hdr->route_type);
 		}
 		break;
 	default:
 		/* Unknown command code or not supported currently */
+		pr_debug("%s: Unsupported Command Code: %d", __func__,
+			 ctrl_hdr->command_code);
 		break;
 	}
 }
diff --git a/include/linux/aspeed-mctp.h b/include/linux/aspeed-mctp.h
index 503b11aff9ff..48628d2abfe2 100644
--- a/include/linux/aspeed-mctp.h
+++ b/include/linux/aspeed-mctp.h
@@ -29,7 +29,7 @@ struct mctp_protocol_hdr {
 } __packed;
 
 #define PCIE_VDM_HDR_SIZE 16
-#define MCTP_BTU_SIZE 64
+#define MCTP_BTU_SIZE 256
 /* The MTU of the ASPEED MCTP can be 64/128/256 */
 #define ASPEED_MCTP_MTU MCTP_BTU_SIZE
 #define PCIE_VDM_DATA_SIZE_DW (ASPEED_MCTP_MTU / 4)
diff --git a/net/mctp/route.c b/net/mctp/route.c
index cf5ae2adf1df..b6befb1a2c0a 100644
--- a/net/mctp/route.c
+++ b/net/mctp/route.c
@@ -16,6 +16,8 @@
 #include <linux/netdevice.h>
 #include <linux/rtnetlink.h>
 #include <linux/skbuff.h>
+#include <linux/dynamic_debug.h>
+#include <linux/printk.h>
 
 #include <uapi/linux/if_arp.h>
 
@@ -1174,8 +1176,15 @@ static int mctp_pkttype_receive(struct sk_buff *skb, struct net_device *dev,
 	if (!rt && mh->dest == MCTP_ADDR_NULL && skb->pkt_type == PACKET_HOST)
 		rt = mctp_route_lookup_null(net, dev);
 
-	if (!rt)
+	if (!rt && mh->dest == MCTP_ADDR_ANY && skb->pkt_type == PACKET_HOST) {
+		pr_debug("%s: Broadcasting", __func__);
+		rt = mctp_route_lookup_null(net, dev);
+	}
+
+	if (!rt) {
+		pr_err("%s: Drop the packet due to no route.", __func__);
 		goto err_drop;
+	}
 
 	rt->output(rt, skb);
 	mctp_route_release(rt);
-- 
2.45.2

